cpp/src/arrow/types/construct.cc

```diff
@@ -71,6 +72,20 @@ Status MakeBuilder(MemoryPool* pool, const std::shared_ptr<DataType>& type,
   }
 }
 
+Status MakeStructBuilder(MemoryPool* pool, const std::shared_ptr<DataType>& type,
```

Comment:
I think this was bad advice for the constructor.  Requiring allocation in the constructor is technically problematic because we don't allow exceptions.  Requiring type here is correct, and anything requiring an allocation should be part of a static factory method (which doesn't need to be done as part of this PR).  I violated this in some of my code, and need to go back and clean it up.